From 184f5f37752623f2c473dcda9e5044ac666141c5 Mon Sep 17 00:00:00 2001
From: Ramneek Mehresh <ramneek.mehresh@freescale.com>
Date: Tue, 24 Mar 2015 17:53:15 +0530
Subject: [PATCH 727/741] usb: pm: Fix usb pm support for deep-sleep

commit 980ba5997ebaca4905a00060a9ec81946bbad95c from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Make sure that saving and restoration of usb phy
registers only happen in case of deep-sleep, and
not for any other feature like sleep, etc. Also
export pm_suspend_state() and set_pm_suspend_state()

Signed-off-by: Ramneek Mehresh <ramneek.mehresh@freescale.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/usb/host/ehci-fsl.c | 32 ++++++++++++++++++++++++++------
 1 file changed, 26 insertions(+), 6 deletions(-)

diff --git a/drivers/usb/host/ehci-fsl.c b/drivers/usb/host/ehci-fsl.c
index cacdadbacc03..bb76c44955b9 100644
--- a/drivers/usb/host/ehci-fsl.c
+++ b/drivers/usb/host/ehci-fsl.c
@@ -26,6 +26,10 @@
 #include <linux/of_platform.h>
 #include <linux/io.h>
 
+#ifdef CONFIG_PM
+#include <linux/suspend.h>
+#endif
+
 #include "ehci.h"
 #include "ehci-fsl.h"
 
@@ -538,6 +542,7 @@ static int ehci_fsl_setup(struct usb_hcd *hcd)
 #ifdef CONFIG_PM
 void __iomem *phy_reg;
 
+#ifdef CONFIG_PPC
 /* save usb registers */
 static int ehci_fsl_save_context(struct usb_hcd *hcd)
 {
@@ -583,6 +588,7 @@ static int ehci_fsl_restore_context(struct usb_hcd *hcd)
 
 	return 0;
 }
+#endif
 
 #ifdef CONFIG_PPC_MPC512x
 static int ehci_fsl_mpc512x_drv_suspend(struct device *dev)
@@ -734,13 +740,20 @@ static int ehci_fsl_drv_suspend(struct device *dev)
 	struct usb_hcd *hcd = dev_get_drvdata(dev);
 	struct ehci_fsl *ehci_fsl = hcd_to_ehci_fsl(hcd);
 	void __iomem *non_ehci = hcd->regs;
-
-	ehci_fsl_save_context(hcd);
-
 #if defined(CONFIG_FSL_USB2_OTG) || defined(CONFIG_FSL_USB2_OTG_MODULE)
 	struct usb_bus host = hcd->self;
 #endif
 
+#ifdef CONFIG_PPC
+	suspend_state_t pm_state;
+	/* FIXME:Need to port fsl_pm.h before enable below code. */
+	/* pm_state = pm_suspend_state(); */
+	pm_state = PM_SUSPEND_MEM;
+
+if (pm_state == PM_SUSPEND_MEM)
+	ehci_fsl_save_context(hcd);
+#endif
+
 	if (of_device_is_compatible(dev->parent->of_node,
 				    "fsl,mpc5121-usb2-dr")) {
 		return ehci_fsl_mpc512x_drv_suspend(dev);
@@ -774,13 +787,20 @@ static int ehci_fsl_drv_resume(struct device *dev)
 	struct ehci_fsl *ehci_fsl = hcd_to_ehci_fsl(hcd);
 	struct ehci_hcd *ehci = hcd_to_ehci(hcd);
 	void __iomem *non_ehci = hcd->regs;
-
-	ehci_fsl_restore_context(hcd);
-
 #if defined(CONFIG_FSL_USB2_OTG) || defined(CONFIG_FSL_USB2_OTG_MODULE)
 	struct usb_bus host = hcd->self;
 #endif
 
+#ifdef CONFIG_PPC
+	suspend_state_t pm_state;
+	/* FIXME:Need to port fsl_pm.h before enable below code.*/
+	/* pm_state = pm_suspend_state(); */
+	pm_state = PM_SUSPEND_MEM;
+
+if (pm_state == PM_SUSPEND_MEM)
+	ehci_fsl_restore_context(hcd);
+#endif
+
 	if (of_device_is_compatible(dev->parent->of_node,
 				    "fsl,mpc5121-usb2-dr")) {
 		return ehci_fsl_mpc512x_drv_resume(dev);
-- 
2.17.1

